server/src/main/java/org/elasticsearch/action/admin/cluster/stats/TransportClusterStatsAction.java

```diff
@@ -376,5 +439,85 @@ AnalysisStats analysisStats() {
         ClusterSnapshotStats clusterSnapshotStats() {
             return clusterSnapshotStats;
         }
+
+        public Map<String, RemoteClusterStats> getRemoteStats() {
+            return remoteStats;
+        }
+    }
+
+    private static boolean doRemotes(ClusterStatsRequest request) {
+        return CCS_TELEMETRY_FEATURE_FLAG.isEnabled() && request.doRemotes();
+    }
+
+    private class RemoteStatsFanout extends CancellableFanOut<String, RemoteClusterStatsResponse, Map<String, RemoteClusterStats>> {
+        private final Map<String, RemoteClusterStatsResponse> responses = new ConcurrentHashMap<>();
+        private final Executor requestExecutor;
+        private final Task task;
+        private final TaskId taskId;
+        private final Collection<String> remotes;
+
+        RemoteStatsFanout(Task task, Executor requestExecutor, Collection<String> remotes) {
+            this.task = task;
+            this.requestExecutor = requestExecutor;
+            this.taskId = new TaskId(clusterService.getNodeName(), task.getId());
+            this.remotes = remotes;
+        }
+
+        @Override
+        protected void sendItemRequest(String clusterAlias, ActionListener<RemoteClusterStatsResponse> listener) {
+            var remoteRequest = new RemoteClusterStatsRequest();
+            var remoteClusterClient = remoteClusterService.getRemoteClusterClient(
+                clusterAlias,
+                requestExecutor,
+                RemoteClusterService.DisconnectedStrategy.RECONNECT_IF_DISCONNECTED
+            );
+            remoteRequest.setParentTask(taskId);
+            remoteClusterClient.execute(TransportRemoteClusterStatsAction.REMOTE_TYPE, remoteRequest, listener);
+        }
+
+        @Override
+        protected void onItemResponse(String clusterAlias, RemoteClusterStatsResponse response) {
+            if (isCancelled() == false) {
+                responses.put(clusterAlias, response);
+            }
+        }
+
+        @Override
+        protected void onItemFailure(String clusterAlias, Exception e) {
+            logger.warn("Failed to get remote cluster stats for [{}]: {}", clusterAlias, e);
+        }
+
+        private boolean isCancelled() {
+            return task instanceof CancellableTask cancellableTask && cancellableTask.isCancelled();
+        }
+
+        void start(ActionListener<Map<String, RemoteClusterStats>> listener) {
+            super.run(task, remotes.iterator(), listener);
+        }
+
+        @Override
+        protected Map<String, RemoteClusterStats> onCompletion() {
+            if (isCancelled()) {
+                return Map.of();
+            }
+
+            Map<String, RemoteClusterStats> remoteClustersStats = new HashMap<>();
+
+            for (String clusterAlias : remotes) {
+                RemoteClusterConnection remoteConnection = remoteClusterService.getRemoteClusterConnection(clusterAlias);
```

Comment:
I'd rather we did this in `onItemResponse`, and it needs protection for the case where the collection of remotes changed underneath us.

Reply:
We can only do part of it on `onItemResponse`. The point of this cycle is that the cluster data has two parts - one which is always present (that's why we go over whole `remotes` list) and one that is derived from the remote response. In the case that remote fails, we don't get the remote response part, but we still need to do the other part, so every remote, even unavailable one, will have an item in the resulting map. We could reverse it, and fill up the map first and then fill in the response part, but that would require to make the RemoteClusterStats object mutable, since they'd still need to be assembled from the two parts, and I like that way less.

Reply:
The remotes list can't change because we keep it in remotes? Unless `getRegisteredRemoteClusterNames` returns something that is not persistent? I will verify it. If it is, then we may have no choice but to go with mutable `RemoteClusterStats`...

Reply:
Sure but we can handle the missing case in `onItemFailure`.

Holding onto `remotes` is not sufficient here, `remoteClusterService.getRemoteClusterConnection()` might still throw if a remote with the given name no longer exists.

Reply:
OK I rewrote it so that it doesn't keep `remotes` anymore. I am not sure what to do though if `getRemoteClusterConnection` throws - it should be extremely unlikely but I guess theoretically possible. Unfortunately, `remoteClusterService` doesn't have any API to iterate over connection list with names. While it has `remoteClusters` map which is what we need, it does not export it. If I add a getter for this map, I could likely avoid this case - even if it's pretty unlikely.

Reply:
Actually, I am wrong - since it's a `keySet()`, the `getRemoteClusterConnection` should be fine now.